File: frontend/src/api/pet.ts
import api from './client';
import { Pet, PetType, PetGender } from '@/types';

interface CreatePetRequest {
  household_id: string;
  name: string;
  pet_type: PetType;
  breed?: string;
  date_of_birth?: string;
  gender?: PetGender;
  microchip_number?: string;
  vet_name?: string;
  vet_phone?: string;
  vet_address?: string;
  insurance_policy_id?: string;
  notes?: string;
}

interface UpdatePetRequest {
  name?: string;
  pet_type?: PetType;
  breed?: string;
  date_of_birth?: string;
  gender?: PetGender;
  microchip_number?: string;
  vet_name?: string;
  vet_phone?: string;
  vet_address?: string;
  insurance_policy_id?: string;
  notes?: string;
}

export const petApi = {
  list: async (householdId?: string): Promise<Pet[]> => {
    const params = householdId ? `?household_id=${householdId}` : '';
    const response = await api.get<Pet[]>(`/pets${params}`);
    return response.data;
  },

  get: async (id: string): Promise<Pet> => {
    const response = await api.get<Pet>(`/pets/${id}`);
    return response.data;
  },

  create: async (data: CreatePetRequest): Promise<Pet> => {
    const response = await api.post<Pet>('/pets', data);
    return response.data;
  },

  update: async (id: string, data: UpdatePetRequest): Promise<Pet> => {
    const response = await api.put<Pet>(`/pets/${id}`, data);
    return response.data;
  },

  delete: async (id: string): Promise<void> => {
    await api.delete(`/pets/${id}`);
  },
};

File: api/internal/models/pet.go
package models

import (
	"time"

	"github.com/google/uuid"
)

// Pet types
const (
	PetTypeDog     = "DOG"
	PetTypeCat     = "CAT"
	PetTypeBird    = "BIRD"
	PetTypeFish    = "FISH"
	PetTypeReptile = "REPTILE"
	PetTypeRabbit  = "RABBIT"
	PetTypeHamster = "HAMSTER"
	PetTypeOther   = "OTHER"
)

// Pet genders
const (
	PetGenderMale    = "MALE"
	PetGenderFemale  = "FEMALE"
	PetGenderUnknown = "UNKNOWN"
)

// Pet represents a pet owned by the household
type Pet struct {
	ID                uuid.UUID  `json:"id"`
	HouseholdID       uuid.UUID  `json:"household_id"`
	Name              string     `json:"name"`
	PetType           string     `json:"pet_type"`
	Breed             string     `json:"breed,omitempty"`
	DateOfBirth       *time.Time `json:"date_of_birth,omitempty"`
	Gender            string     `json:"gender,omitempty"`
	MicrochipNumber   string     `json:"microchip_number,omitempty"`
	VetName           string     `json:"vet_name,omitempty"`
	VetPhone          string     `json:"vet_phone,omitempty"`
	VetAddress        string     `json:"vet_address,omitempty"`
	InsurancePolicyID *uuid.UUID `json:"insurance_policy_id,omitempty"`
	Notes             string     `json:"notes,omitempty"`
	CreatedAt         time.Time  `json:"created_at"`
	UpdatedAt         time.Time  `json:"updated_at"`
	// Computed fields
	Age               *int       `json:"age,omitempty"`
}

// CalculateAge calculates the pet's age in years
func (p *Pet) CalculateAge() {
	if p.DateOfBirth == nil {
		p.Age = nil
		return
	}
	now := time.Now()
	years := now.Year() - p.DateOfBirth.Year()
	if now.YearDay() < p.DateOfBirth.YearDay() {
		years--
	}
	p.Age = &years
}

// ValidPetTypes returns all valid pet types
func ValidPetTypes() []string {
	return []string{
		PetTypeDog,
		PetTypeCat,
		PetTypeBird,
		PetTypeFish,
		PetTypeReptile,
		PetTypeRabbit,
		PetTypeHamster,
		PetTypeOther,
	}
}

// ValidPetGenders returns all valid pet genders
func ValidPetGenders() []string {
	return []string{
		PetGenderMale,
		PetGenderFemale,
		PetGenderUnknown,
	}
}
